### DISCUSSION TITLE
Wrapping Prisma calls with dynamic arguments

### DISCUSSION THREAD

**Discussion by nehalist** [prisma/prisma#19501] [Q&A]
I do have a mono repo and want the data layer to be a completely separate package, so I'm currently extracting all data-related stuff to this package. Unfortunately types do cause some issues for me:

```
(async () => {
  async function getTasks<T extends Prisma.TaskFindManyArgs>(
    args?: Prisma.SelectSubset<T, Prisma.TaskFindManyArgs>
  ) {
    return prisma.task.findMany<T>({
      where: {
        status: TaskStatus.Pending,
      },
      ...args,
    });
  }

  function foo(tasks: TaskWithUserAndItem[]) {
    return tasks;
  }

  const tasks = await getTasks({ include: { user: true, item: true } });

  foo(tasks); // doesn't work, since tasks is `Task[]` and not `TaskWithUserAndItem[]`
})();
```
TaskWithUserAndItem is a custom type that requires the Task to include User and Item (by adding `include: { user: true, item, true }` to findMany).

Unfortunately, if TypeScript strict is mode enabled the return of getTasks is invalid according to TypeScript with the following error:

TS2345: Argument of type '{ select?: T["select"] | undefined; include?: T["include"] | undefined; where: T["where"] | { status: "Pending"; }; orderBy?: T["orderBy"] | undefined; cursor?: T["cursor"] | undefined; take?: T["take"] | undefined; skip?: T["skip"] | undefined; distinct?: T["distinct"] | undefined; }' is not assignable to parameter of type 'SelectSubset<T, TaskFindManyArgs>'.   Type '{ select?: T["select"] | undefined; include?: T["include"] | undefined; where: T["where"] | { status: "Pending"; }; orderBy?: T["orderBy"] | undefined; cursor?: T["cursor"] | undefined; take?: T["take"] | undefined; skip?: T["skip"] | undefined; distinct?: T["distinct"] | undefined; }' is not assignable to type '{ [key in keyof T]: key extends keyof TaskFindManyArgs ? T[key] : never; }'.
Disabling strict mode does work, but obviously can't be the correct way.

How to achieve this?

**Comment by nurul3101** [comment]
Hello @nehalist 👋 

> TaskWithUserAndItem is a custom type that requires the Task to include User and Item (by adding include: { user: true, item, true }

You can create a custom type that has tasks along with user and item.
Did you had a chance to look at this [section](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#problem-using-variations-of-the-generated-model-type) in which we use `Prisma.validator` API to create a custom type?

**Reply by nehalist** [reply]
My custom type is already using this API:

```typescript
export const TaskWithUser = Prisma.validator<Prisma.TaskArgs>()({
  include: {
    user: true,
  },
});
export type TaskWithUser = Prisma.TaskGetPayload<typeof TaskWithUser>;

export const TaskWithItem = Prisma.validator<Prisma.TaskArgs>()({
  include: {
    item: true,
  },
});
export type TaskWithItem = Prisma.TaskGetPayload<typeof TaskWithItem>;

export type TaskWithUserAndItem = TaskWithUser & TaskWithItem;
```

**Reply by nurul3101** [reply]
Okay, would it be possible for you to provide a minimal reproduction so that I can try to figure out the cause for the type error?

**Reply by nurul3101** [reply]
Just checking to see if you had a chance to look at my last response.

**Reply by nehalist** [reply]
Hey, sorry I haven't replied so far - been on vacation this week. I'll provide a repo asap :)

**Reply by nurul3101** [reply]
Thanks for getting back and letting us know, a reproduction would definitely be helpful.

**Reply by nehalist** [reply]
https://github.com/nehalist/prisma-func-wrapper

See `index.ts`

`npm start` works, but disabling `strict` mode in `tsconfig.json` causes Type errors in the `getTasks` function

**Reply by IzioDev** [reply]
note that this doesn't work with `5.1.0` anymore

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
